Content.Shared/Xenoarchaeology/Artifact/SharedXenoArtifactSystem.Graph.cs

```diff
@@ -0,0 +1,461 @@
+    }
+
+    public bool TryGetNode(Entity<XenoArtifactComponent?> ent, int index, [NotNullWhen(true)] out Entity<XenoArtifactNodeComponent>? node)
+    {
+        node = null;
+        if (!Resolve(ent, ref ent.Comp))
+            return false;
+
+        if (index < 0 || index >= ent.Comp.NodeVertices.Length)
+            return false;
+
+        if (ent.Comp.NodeVertices[index] is { } netUid && GetEntity(netUid) is var uid)
+            node = (uid, XenoArtifactNode(uid));
+
+        return node != null;
+    }
+
+    /// <summary>
+    /// Gets the index of the first empty spot in the NodeVertices array.
+    /// If there is none, resizes both arrays and returns the new index.
+    /// </summary>
+    public int GetFreeNodeIndex(Entity<XenoArtifactComponent> ent)
+    {
+        var length = ent.Comp.NodeVertices.Length;
+        for (var i = 0; i < length; i++)
+        {
+            if (ent.Comp.NodeVertices[i] == null)
+                return i;
+        }
+
+        ResizeNodeGraph(ent, length + 1);
+        return length;
+    }
+
+    public IEnumerable<Entity<XenoArtifactNodeComponent>> GetAllNodes(Entity<XenoArtifactComponent> ent)
+    {
+        foreach (var netNode in ent.Comp.NodeVertices)
+        {
+            if (GetEntity(netNode) is { } node)
+                yield return (node, XenoArtifactNode(node));
+        }
+    }
+
+    public IEnumerable<int> GetAllNodeIndices(Entity<XenoArtifactComponent> ent)
+    {
+        for (var i = 0; i < ent.Comp.NodeVertices.Length; i++)
+        {
+            if (ent.Comp.NodeVertices[i] is not null)
+                yield return i;
+        }
+    }
+
+    public bool AddEdge(Entity<XenoArtifactComponent?> ent, EntityUid from, EntityUid to, bool dirty = true)
+    {
+        if (!Resolve(ent, ref ent.Comp))
+            return false;
+
+        if (!TryGetIndex(ent, from, out var fromIdx) ||
+            !TryGetIndex(ent, to, out var toIdx))
+            return false;
+
+        return AddEdge(ent, fromIdx.Value, toIdx.Value, dirty: dirty);
+    }
+
+    public bool AddEdge(Entity<XenoArtifactComponent?> ent, int fromIdx, int toIdx, bool dirty = true)
+    {
+        if (!Resolve(ent, ref ent.Comp))
+            return false;
+
+        DebugTools.Assert(fromIdx >= 0 && fromIdx < ent.Comp.NodeVertices.Length, $"fromIdx is out of bounds for fromIdx {fromIdx}");
+        DebugTools.Assert(toIdx >= 0 && toIdx < ent.Comp.NodeVertices.Length, $"toIdx is out of bounds for toIdx {toIdx}");
+
+        if (ent.Comp.NodeAdjacencyMatrix[fromIdx][toIdx])
+            return false; //Edge already exists
+
+        // TODO: add a safety check to prohibit cyclic paths.
+
+        ent.Comp.NodeAdjacencyMatrix[fromIdx][toIdx] = true;
+        if (dirty)
+        {
+            RebuildXenoArtifactMetaData(ent);
+        }
+
+        CancelUnlockingOnGraphStructureChange((ent, ent.Comp));
```

Comment:
You've added a bunch of calls to `CancelUnlockingOnGraphStructureChange()` throughout this file and i think they'd be better centralized inside of `RebuildXenoArtifactMetaData()`. That function exists to update things like segment metadata and as such gets called once when the xenoartifact's graph is modified in some way.

In that sense, it's better to place the calls there because you greatly reduce the number of calls. All the different functions in this system get called 10s of times in a single graph generation, which means that you are going to attempt to cancel the unlock (which has a tryComp call) a bunch of times for no reason.

Reply:
While i don't like the idea to place what can be called 'business logic' into the 'infrastructure method'... it is very nice to remove those 'extra' calls from generation process, and make code less intertwined.